ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java

```diff
@@ -76,11 +79,74 @@ public void testKsqlResource() {
 
 
   @Test
-  public void testStreamQuery() {
+  public void testStreamRowFromServer() throws InterruptedException {
+    MockStreamedQueryResource sqr = mockApplication.getStreamedQueryResource();
     RestResponse<KsqlRestClient.QueryStream> queryResponse = ksqlRestClient.makeQueryRequest
-        ("Select *");
+            ("Select *");
     Assert.assertNotNull(queryResponse);
     Assert.assertTrue(queryResponse.isSuccessful());
+
+    // Get the stream writer from the mock server and load it up with a row
+    List<MockStreamedQueryResource.TestStreamWriter> writers = sqr.getWriters();
+    Assert.assertEquals(1, writers.size());
+    MockStreamedQueryResource.TestStreamWriter writer = writers.get(0);
+    try {
+      writer.enq("hello");
+
+      // Try and receive the row. Do this from another thread to avoid blocking indefinitely
+      KsqlRestClient.QueryStream queryStream = queryResponse.getResponse();
+      Thread t = new Thread(() -> queryStream.hasNext());
+      t.setDaemon(true);
+      t.start();
+      t.join(10000);
+      Assert.assertFalse(t.isAlive());
+      Assert.assertTrue(queryStream.hasNext());
+
+      StreamedRow sr = queryStream.next();
+      Assert.assertNotNull(sr);
+      GenericRow row = sr.getRow();
+      Assert.assertEquals(1, row.getColumns().
+              size());
+      Assert.assertEquals("hello", row.getColumns().
+              get(0));
+    } finally {
+      writer.finished();
+    }
+  }
+
+  @Test
+  public void shouldInterruptScannerOnClose() throws InterruptedException {
+    MockStreamedQueryResource sqr = mockApplication.getStreamedQueryResource();
+    RestResponse<KsqlRestClient.QueryStream> queryResponse = ksqlRestClient.makeQueryRequest
+            ("Select *");
+    Assert.assertNotNull(queryResponse);
+    Assert.assertTrue(queryResponse.isSuccessful());
+
+    List<MockStreamedQueryResource.TestStreamWriter> writers = sqr.getWriters();
+    Assert.assertEquals(1, writers.size());
+    try {
+      // Try and receive a row. This will block since there is no data to return
+      KsqlRestClient.QueryStream queryStream = queryResponse.getResponse();
+      CountDownLatch threw = new CountDownLatch(1);
+      Thread t = new Thread(() -> {
+        try {
+          queryStream.hasNext();
+        } catch (IllegalStateException e) {
+          threw.countDown();
+        }
+      });
+      t.setDaemon(true);
+      t.start();
+
+      // Let the thread run and then close the stream. Verify that it was interrupted
+      Thread.sleep(100);
```

Comment:
rather than sleeping wait on the latch, i.e,
`assertTrue(threw.await(100))`

Reply:
The intent behind the sleep was to give the thread a chance to run and block so that it gets interrupted when close is called. That probably won't happen if I call close immediately and then wait on the latch.

Reply:
actually what i think you should do here is immediately close the `QueryStream` after line 129. Then the first call to `queryStream.hasNext()` will throw the expected exception. You can remove the sleep and just wait on the latch

Reply:
I understand. My objection is that I want the query thread to actually sleep, so that I can verify that it behaves correctly after being interrupted. Implementing the test the way you are suggesting will result in the query thread throwing the exception before it sleeps. That said, I'm not a huge fan of timing-based tests. But I don't see a better way of doing it other than instrumenting QueryStream just for the sake of testing this case.